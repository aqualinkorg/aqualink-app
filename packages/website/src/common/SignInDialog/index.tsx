import React, {
  BaseSyntheticEvent,
  useCallback,
  useEffect,
  useState,
} from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Dialog,
  Card,
  CardHeader,
  CardContent,
  Grid,
  Typography,
  IconButton,
  TextField,
  Button,
  LinearProgress,
  Collapse,
} from "@material-ui/core";
import { Alert } from "@material-ui/lab";
import CloseIcon from "@material-ui/icons/Close";
import { useForm } from "react-hook-form";
import { useDispatch, useSelector } from "react-redux";

import {
  resetPassword,
  signInUser,
  userInfoSelector,
  userLoadingSelector,
  userErrorSelector,
} from "../../store/User/userSlice";
import { UserSignInParams } from "../../store/User/types";
import dialogStyles from "../styles/dialogStyles";

const SignInDialog = ({
  open,
  handleRegisterOpen,
  handleSignInOpen,
  classes,
}: SignInDialogProps) => {
  const dispatch = useDispatch();
  const user = useSelector(userInfoSelector);
  const loading = useSelector(userLoadingSelector);
  const error = useSelector(userErrorSelector);
  const [errorAlertOpen, setErrorAlertOpen] = useState<boolean>(false);
  const [passwordResetEmail, setPasswordResetEmail] = useState<string>("");
  const { register, errors, handleSubmit } = useForm({
    reValidateMode: "onSubmit",
  });

  const onSubmit = useCallback(
    (
      data: any,
      event?: BaseSyntheticEvent<object, HTMLElement, HTMLElement>
    ) => {
      if (event) {
        event.preventDefault();
      }
      const registerInfo: UserSignInParams = {
        email: data.emailAddress,
        password: data.password,
      };
      dispatch(signInUser(registerInfo));
    },
    [dispatch]
  );

  useEffect(() => {
    if (user) {
      handleSignInOpen(false);
    }
    if (error) {
      setErrorAlertOpen(true);
    }
  }, [user, handleSignInOpen, error]);

  const onResetPassword = useCallback(
    (
      data: any,
      event?: BaseSyntheticEvent<object, HTMLElement, HTMLElement>
    ) => {
      if (event) {
        event.preventDefault();
      }
      dispatch(resetPassword({ email: data.emailAddress }));
      setPasswordResetEmail(data.emailAddress);
    },
    [dispatch]
  );

  return (
    <Dialog open={open} maxWidth="md">
      <Card>
        <CardHeader
          className={classes.dialogHeader}
          title={
            <Grid container justify="center" item xs={12}>
              <Grid
                container
                alignItems="center"
                justify="space-around"
                item
                xs={12}
              >
                <Grid container item xs={6}>
                  <Typography variant="h4">Aqua</Typography>
                  <Typography
                    className={classes.dialogHeaderSecondPart}
                    variant="h4"
                  >
                    link
                  </Typography>
                </Grid>
                <Grid container justify="flex-end" item xs={1}>
                  <IconButton
                    className={classes.closeButton}
                    size="small"
                    onClick={() => handleSignInOpen(false)}
                  >
                    <CloseIcon />
                  </IconButton>
                </Grid>
              </Grid>
            </Grid>
          }
        />
        {loading && <LinearProgress />}
        {error && (
          <Collapse in={errorAlertOpen}>
            <Alert
              severity="error"
              action={
                <IconButton
                  aria-label="close"
                  color="inherit"
                  size="small"
                  onClick={() => {
                    setErrorAlertOpen(false);
                  }}
                >
                  <CloseIcon fontSize="inherit" />
                </IconButton>
              }
            >
              {error}
            </Alert>
          </Collapse>
        )}
        <Collapse in={passwordResetEmail !== ""}>
          <Alert
            severity="success"
            action={
              <IconButton
                aria-label="close"
                color="inherit"
                size="small"
                onClick={() => {
                  setPasswordResetEmail("");
                }}
              >
                <CloseIcon fontSize="inherit" />
              </IconButton>
            }
          >
            {`Password reset email sent to ${passwordResetEmail}.`}
          </Alert>
        </Collapse>
        <CardContent>
          <Grid container justify="center" item xs={12}>
            <Grid className={classes.dialogContentTitle} container item xs={10}>
              <Grid item>
                <Typography variant="h5" color="textSecondary">
                  Sign In
                </Typography>
              </Grid>
            </Grid>
            <Grid container item xs={10}>
              <form className={classes.form} onSubmit={handleSubmit(onSubmit)}>
                <Grid className={classes.textFieldWrapper} item xs={12}>
                  <Typography className={classes.formText} variant="subtitle2">
                    or login with email address
                  </Typography>
                  <TextField
                    id="emailAddress"
                    name="emailAddress"
                    placeholder="Email Address"
                    helperText={
                      errors.emailAddress ? errors.emailAddress.message : ""
                    }
                    label="Email Address"
                    inputRef={register({
                      required: "This is a required field",
                      pattern: {
                        value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                        message: "Invalid email address",
                      },
                    })}
                    error={!!errors.emailAddress}
                    inputProps={{ className: classes.textField }}
                    fullWidth
                    variant="outlined"
                  />
                </Grid>
                <Grid className={classes.textFieldWrapper} item xs={12}>
                  <TextField
                    id="password"
                    name="password"
                    type="password"
                    placeholder="Password"
                    helperText={errors.password ? errors.password.message : ""}
                    label="Password"
                    inputRef={register({
                      required: "This is a required field",
                    })}
                    error={!!errors.password}
                    inputProps={{ className: classes.textField }}
                    fullWidth
                    variant="outlined"
                  />
                </Grid>
                <Grid container item xs={12}>
                  <Button
                    className={classes.forgotPasswordButton}
                    onClick={handleSubmit(onResetPassword)}
                  >
                    <Typography variant="subtitle2" color="textSecondary">
                      Forgot your password?
                    </Typography>
                  </Button>
                </Grid>
                <Grid className={classes.button} item xs={12}>
                  <Button
                    fullWidth
                    type="submit"
                    color="primary"
                    variant="contained"
                    size="large"
                  >
                    SIGN IN
                  </Button>
                </Grid>
                <Grid container item xs={12}>
                  <Typography
                    className={classes.formText}
                    variant="subtitle1"
                    color="textSecondary"
                  >
                    Don&#39;t have an account?{" "}
                    <Button
                      onClick={() => {
                        handleRegisterOpen(true);
                        handleSignInOpen(false);
                      }}
                      color="primary"
                    >
                      REGISTER
                    </Button>
                  </Typography>
                </Grid>
              </form>
            </Grid>
          </Grid>
        </CardContent>
      </Card>
    </Dialog>
  );
};

const styles = () =>
  createStyles({
    ...dialogStyles,
    formText: {
      ...dialogStyles.formText,
      marginBottom: "1rem",
    },
    forgotPasswordButton: {
      color: "inherit",
      textDecoration: "none",
    },
  });

interface SignInDialogIncomingProps {
  open: boolean;
  handleRegisterOpen: (open: boolean) => void;
  handleSignInOpen: (open: boolean) => void;
}

type SignInDialogProps = SignInDialogIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(SignInDialog);
