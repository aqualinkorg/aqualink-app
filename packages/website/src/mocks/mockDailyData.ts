import { DailyData } from "../store/Reefs/types";

const now = new Date();
const minutesAgo = 5;
const dailyDataDate = new Date(
  now.getTime() - minutesAgo * 60000
).toISOString();

export const mockDailyData: DailyData = {
  id: 1,
  date: dailyDataDate,
  minBottomTemperature: 25,
  maxBottomTemperature: 27,
  avgBottomTemperature: 26,
  degreeHeatingDays: 29,
  surfaceTemperature: 36,
  satelliteTemperature: 20,
  minWaveHeight: 1,
  maxWaveHeight: 3,
  avgWaveHeight: 2,
  waveDirection: 136,
  wavePeriod: 15,
  minWindSpeed: 3,
  maxWindSpeed: 5,
  avgWindSpeed: 4,
  windDirection: 96,
  weeklyAlertLevel: 3,
};
