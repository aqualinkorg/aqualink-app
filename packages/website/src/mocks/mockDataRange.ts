import { TimeSeriesDataRange } from 'store/Sites/types';

export const mockDataRange: TimeSeriesDataRange = {
  bottomTemperature: [
    {
      type: 'hobo',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  cholorophyllConcentration: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  odoConcentration: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  ph: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  salinity: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
    {
      type: 'hui',
      depth: null,
      data: [
        {
          minDate: '2022-06-11T00:00:00.000Z',
          maxDate: '2022-06-14T23:45:00.000Z',
        },
      ],
    },
  ],
  turbidity: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
};
