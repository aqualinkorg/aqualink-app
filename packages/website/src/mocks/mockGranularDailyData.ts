import { DailyData } from 'store/Sites/types';

export const mockGranularDailyData: DailyData[] = [
  {
    id: 3,
    date: '2022-06-27T02:59:59.999Z',
    minBottomTemperature: 0,
    maxBottomTemperature: 0,
    avgBottomTemperature: 0,
    degreeHeatingDays: 0,
    topTemperature: 0,
    satelliteTemperature: 24.219252436179815,
    minWaveHeight: 0.17217956015502844,
    maxWaveHeight: 0.4220004534334179,
    avgWaveHeight: 0.28517965469411594,
    waveMeanDirection: 219,
    waveMeanPeriod: 2,
    minWindSpeed: 1.0233644143488965,
    maxWindSpeed: 5.558413659400355,
    avgWindSpeed: 2.9796803902255204,
    windDirection: 218,
    weeklyAlertLevel: 0,
  },
  {
    id: 2,
    date: '2022-06-26T02:59:59.999Z',
    minBottomTemperature: 0,
    maxBottomTemperature: 0,
    avgBottomTemperature: 0,
    degreeHeatingDays: 0,
    topTemperature: 0,
    satelliteTemperature: 24.229605847456323,
    minWaveHeight: 0.06706502244580449,
    maxWaveHeight: 0.40127747690701104,
    avgWaveHeight: 0.20286431769935753,
    waveMeanDirection: 186,
    waveMeanPeriod: 2,
    minWindSpeed: 1.0683246478923132,
    maxWindSpeed: 3.8642400151240097,
    avgWindSpeed: 2.4470019134494967,
    windDirection: 238,
    weeklyAlertLevel: 0,
  },
  {
    id: 1,
    date: '2022-06-25T02:59:59.999Z',
    minBottomTemperature: 0,
    maxBottomTemperature: 0,
    avgBottomTemperature: 0,
    degreeHeatingDays: 0,
    topTemperature: 0,
    satelliteTemperature: 24.00602354214886,
    minWaveHeight: 0.25399888336189147,
    maxWaveHeight: 0.7053866500812253,
    avgWaveHeight: 0.49466218581824045,
    waveMeanDirection: 184,
    waveMeanPeriod: 3,
    minWindSpeed: 2.3608638556613584,
    maxWindSpeed: 6.910529333003823,
    avgWindSpeed: 4.019357707456016,
    windDirection: 125,
    weeklyAlertLevel: 0,
  },
];
