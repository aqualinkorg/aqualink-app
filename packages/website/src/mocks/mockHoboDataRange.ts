import { TimeSeriesDataRange } from "../store/Reefs/types";

export const mockHoboDataRange: TimeSeriesDataRange = {
  hobo: {
    bottomTemperature: [
      {
        minDate: "2017-11-02T14:59:08.000Z",
        maxDate: "2020-02-02T20:59:08.000Z",
      },
    ],
    alert: [],
    dhw: [],
    satelliteTemperature: [],
    topTemperature: [],
    sstAnomaly: [],
    significantWaveHeight: [],
    wavePeakPeriod: [],
    waveMeanDirection: [],
    windSpeed: [],
    windDirection: [],
  },
  spotter: {
    bottomTemperature: [],
    alert: [],
    dhw: [],
    satelliteTemperature: [],
    topTemperature: [],
    sstAnomaly: [],
    significantWaveHeight: [],
    wavePeakPeriod: [],
    waveMeanDirection: [],
    windSpeed: [],
    windDirection: [],
  },
  sofarApi: {
    bottomTemperature: [],
    alert: [],
    dhw: [],
    satelliteTemperature: [],
    topTemperature: [],
    sstAnomaly: [],
    significantWaveHeight: [],
    wavePeakPeriod: [],
    waveMeanDirection: [],
    windSpeed: [],
    windDirection: [],
  },
};
