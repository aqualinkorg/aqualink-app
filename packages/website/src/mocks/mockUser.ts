import { User } from "../store/User/types";

export const mockUser: User = {
  adminLevel: "default",
  administeredReefs: [],
  email: "mail@mail.com",
  firebaseUid: "RaNdOm StRiNg",
  fullName: "Mock User",
  organization: "Some organization",
  token: "RaNdOm StRiNg",
};
