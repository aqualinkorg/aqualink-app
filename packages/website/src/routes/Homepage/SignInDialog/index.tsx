import React, {
  BaseSyntheticEvent,
  useCallback,
  useEffect,
  useState,
} from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Theme,
  Dialog,
  Card,
  CardHeader,
  CardContent,
  Grid,
  Typography,
  IconButton,
  TextField,
  Button,
  LinearProgress,
  Collapse,
} from "@material-ui/core";
import { Alert } from "@material-ui/lab";
import CloseIcon from "@material-ui/icons/Close";
import { useForm } from "react-hook-form";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import {
  signInUser,
  userInfoSelector,
  userLoadingSelector,
  userErrorSelector,
} from "../../../store/User/userSlice";
import { UserSignInParams } from "../../../store/User/types";

const SignInDialog = ({
  open,
  handleRegisterOpen,
  handleSignInOpen,
  classes,
}: SignInDialogProps) => {
  const dispatch = useDispatch();
  const user = useSelector(userInfoSelector);
  const loading = useSelector(userLoadingSelector);
  const error = useSelector(userErrorSelector);
  const [errorAlertOpen, setErrorAlertOpen] = useState<boolean>(false);
  const { register, errors, handleSubmit } = useForm({
    reValidateMode: "onSubmit",
  });

  const onSubmit = useCallback(
    (
      data: any,
      event?: BaseSyntheticEvent<object, HTMLElement, HTMLElement>
    ) => {
      if (event) {
        event.preventDefault();
      }
      const registerInfo: UserSignInParams = {
        email: data.emailAddress,
        password: data.password,
      };
      dispatch(signInUser(registerInfo));
    },
    [dispatch]
  );

  useEffect(() => {
    if (user) {
      handleSignInOpen(false);
    }
    if (error) {
      setErrorAlertOpen(true);
    }
  }, [user, handleSignInOpen, error]);

  return (
    <Dialog open={open}>
      <Card className={classes.root}>
        <CardHeader
          className={classes.dialogHeader}
          title={
            <Grid container justify="flex-end" item xs={12}>
              <Grid
                container
                alignItems="center"
                justify="space-between"
                item
                xs={11}
              >
                <Grid container item xs={4}>
                  <Typography variant="h4">Aqua</Typography>
                  <Typography style={{ color: "#8AC6DE" }} variant="h4">
                    link
                  </Typography>
                </Grid>
                <Grid item xs={1}>
                  <IconButton
                    className={classes.closeButton}
                    size="small"
                    onClick={() => handleSignInOpen(false)}
                  >
                    <CloseIcon />
                  </IconButton>
                </Grid>
              </Grid>
            </Grid>
          }
        />
        {loading && <LinearProgress />}
        {error && (
          <Collapse in={errorAlertOpen}>
            <Alert
              severity="error"
              action={
                <IconButton
                  aria-label="close"
                  color="inherit"
                  size="small"
                  onClick={() => {
                    setErrorAlertOpen(false);
                  }}
                >
                  <CloseIcon fontSize="inherit" />
                </IconButton>
              }
            >
              {error}
            </Alert>
          </Collapse>
        )}
        <CardContent>
          <Grid container justify="center" item xs={12}>
            <Grid style={{ margin: "1rem 0 1rem 0" }} container item xs={10}>
              <Grid item>
                <Typography variant="h5" color="textSecondary">
                  Sign In
                </Typography>
              </Grid>
            </Grid>
            <Grid container item xs={10}>
              <form
                className={classes.signInForm}
                onSubmit={handleSubmit(onSubmit)}
              >
                <Grid className={classes.textFieldWrapper} item xs={12}>
                  <Typography className={classes.formText} variant="subtitle2">
                    or login with email address
                  </Typography>
                  <TextField
                    id="emailAddress"
                    name="emailAddress"
                    placeholder="Email Address"
                    helperText={
                      errors.emailAddress ? errors.emailAddress.message : ""
                    }
                    label="Email Address"
                    inputRef={register({
                      required: "This is a required field",
                      pattern: {
                        value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                        message: "Invalid email address",
                      },
                    })}
                    error={!!errors.emailAddress}
                    inputProps={{ className: classes.textField }}
                    fullWidth
                    variant="outlined"
                  />
                </Grid>
                <Grid className={classes.textFieldWrapper} item xs={12}>
                  <TextField
                    id="password"
                    name="password"
                    type="password"
                    placeholder="Password"
                    helperText={errors.password ? errors.password.message : ""}
                    label="Password"
                    inputRef={register({
                      required: "This is a required field",
                    })}
                    error={!!errors.password}
                    inputProps={{ className: classes.textField }}
                    fullWidth
                    variant="outlined"
                  />
                </Grid>
                <Grid container item xs={12}>
                  <Typography variant="subtitle2" color="textSecondary">
                    <Link
                      style={{ color: "inherit", textDecoration: "none" }}
                      to="/"
                    >
                      Forgot your password?
                    </Link>
                  </Typography>
                </Grid>
                <Grid className={classes.button} item xs={12}>
                  <Button
                    fullWidth
                    type="submit"
                    color="primary"
                    variant="contained"
                    size="large"
                  >
                    SIGN IN
                  </Button>
                </Grid>
                <Grid container item xs={12}>
                  <Typography
                    className={classes.formText}
                    variant="subtitle1"
                    color="textSecondary"
                  >
                    Don&#39;t have an account?{" "}
                    <Button
                      onClick={() => {
                        handleRegisterOpen(true);
                        handleSignInOpen(false);
                      }}
                      color="primary"
                    >
                      REGISTER
                    </Button>
                  </Typography>
                </Grid>
              </form>
            </Grid>
          </Grid>
        </CardContent>
      </Card>
    </Dialog>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    root: {
      height: "70vh",
      width: "30vw",
    },
    closeButton: {
      color: theme.palette.primary.light,
    },
    dialogHeader: {
      backgroundColor: theme.palette.primary.main,
    },
    signInForm: {
      width: "100%",
    },
    textFieldWrapper: {
      margin: "1rem 0 1rem 0",
    },
    textField: {
      color: "black",
    },
    formText: {
      color: theme.palette.grey[500],
      fontWeight: 400,
      marginBottom: "1rem",
    },
    button: {
      margin: "2rem 0 1rem 0",
    },
  });

interface SignInDialogIncomingProps {
  open: boolean;
  handleRegisterOpen: (open: boolean) => void;
  handleSignInOpen: (open: boolean) => void;
}

type SignInDialogProps = SignInDialogIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(SignInDialog);
