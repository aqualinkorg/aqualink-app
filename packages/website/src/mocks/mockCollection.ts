import { CollectionDetails } from "../store/Collection/types";
import { mockReef } from "./mockReef";
import { mockUser } from "./mockUser";

export const mockCollection: CollectionDetails = {
  id: 1,
  isPublic: false,
  name: "Mock Collection",
  user: mockUser,
  reefs: [mockReef],
  reefIds: [1],
};
