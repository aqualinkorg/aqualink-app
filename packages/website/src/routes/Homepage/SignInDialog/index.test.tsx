import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { BrowserRouter as Router } from "react-router-dom";
import { mockUser } from "../../../mocks/mockUser";

import SignInDialog from ".";

const mockStore = configureStore([]);

describe("SignInDialog", () => {
  let element: HTMLElement;
  beforeEach(() => {
    const store = mockStore({
      user: {
        userInfo: mockUser,
        loading: false,
        error: null,
      },
    });

    store.dispatch = jest.fn();

    element = render(
      <Provider store={store}>
        <Router>
          <SignInDialog
            open
            handleRegisterOpen={jest.fn}
            handleSignInOpen={jest.fn}
          />
        </Router>
      </Provider>
    ).container;
  });

  it("should render with given state from Redux store", () => {
    expect(element).toMatchSnapshot();
  });
});
