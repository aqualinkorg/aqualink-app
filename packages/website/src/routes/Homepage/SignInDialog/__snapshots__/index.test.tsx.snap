// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInDialog should render with given state from Redux store 1`] = `
<div
  aria-hidden="true"
/>
`;
